File: examples/python/src/receiver_txt.c
/*
This is an example of a C-based simpl receiver
*/

#include <stdio.h>
#include <stdlib.h>
#include <simpl.h>

typedef struct
	{
	int token;
	int var1;
	int var2;
	int var3;
	} OUT1;
	
typedef struct
	{
	int token;
	float var1;
	} OUT2;
	
typedef struct
	{
	int token;
	int var1;
	char var2[34];
	} OUT3;
	
int main()
{
char *sender;
char mem[1024];
int n;
OUT1 *in1;
OUT2 *in2;
OUT3 *in3;
int *token;

if (name_attach("receiver", NULL) == -1)
	{
	printf("receiver: cannot attach name: %s\n", whatsMyError());
	exit(-1);
	}

while (1)
	{
	n = Receive(&sender, mem, 1024);
	if (n == -1)
		{
		printf("receiver: Receive error: %s\n", whatsMyError());
		continue;
		}
	
	token = (int *)mem;
	
	switch (*token)
		{
		case 10:
			in1 = (OUT1 *)mem;		
			printf("token=%d var1=%d var2=%d var3=%d\n",
				in1->token,
				in1->var1,
				in1->var2,
				in1->var3);
			break;

		case 20:
			in2 = (OUT2 *)mem;		
			printf("token=%d var1=%f\n",
				in2->token,
				in2->var1);
			break;

		case 30:
			in3 = (OUT3 *)mem;		
			printf("token=%d var1=%d var2=%32s\n",
				in3->token,
				in3->var1,
				in3->var2);
			break;

		default:
			printf("unknown token=%d\n", *token);

		}

	Reply(sender, NULL, 0);
	}

return(0);
}

File: examples/python/src/receiver_txt.py
"""
This is a python simpl text-based example receiver
"""

"""
you will need to either
1. export PYTHONPATH=/home/simpl/simpl.2.0.3/python/modules
OR
2. copy psimpl.py to /usr/lib/python1.5/lib-dynload
for importing psimpl
"""

#! /usr/bin/python

import sys
import struct
from psimpl import getBinaryValue, SINT, SFLOAT, SLONG
from simpl import name_attach, Receive, Reply, whatsMyError

# set this program's name
rName = "py_receiver"

# attach a simpl name
retVal = name_attach(rName)
if retVal == -1:
	print "%s: name attach error-%s" %(rName, whatsMyError())
	sys.exit(-1)

# receive and process a message
while 1:
	# receive a message
	messageSize, senderId, message = Receive()

	# check for error
	if messageSize == -1:
		print "%s: receive error-%s" %(rName, whatsMyError())
		sys.exit(-1)
	# is it a non-null message?
	elif messageSize > 0:
		# extract the token from the message
		token = getBinaryValue(message, 0, SINT, "i")

		# react on token
		if token == 10:
			var1 = getBinaryValue(message, SINT, SINT, "i")
			var2 = getBinaryValue(message, 2 * SINT, SINT, "i")
			var3 = getBinaryValue(message, 3 * SINT, SINT, "i")
			print "%s: token=%d var1=%d var2=%d var3=%d" %(rName, token, var1, var2, var3)

		elif token == 20:
			var1 = getBinaryValue(message, SINT, SFLOAT, "f")
			print "%s: token=%d var1=%f" %(rName, token, var1)

		elif token == 30:
			length = getBinaryValue(message, SINT, SFLOAT, "i")
			var1 = getBinaryValue(message, 2 * SINT, length, "s")
			print "%s: token=%d var1=%s" %(rName, token, var1)

		else:
			print "%s: unknown token type=%d" %(rName, token)

	retVal = Reply(senderId, None)
	if retVal == -1:
		print "%s: reply error-%s" %(rName, whatsMyError())
		sys.exit(-1)
